program-runtime/src/loaded_programs.rs

```diff
@@ -258,11 +288,23 @@ impl LoadedPrograms {
         let index = second_level
             .iter()
             .position(|at| at.effective_slot >= entry.effective_slot);
-        if let Some(existing) = index.and_then(|index| second_level.get(index)) {
+        if let Some((existing, entry_index)) =
+            index.and_then(|index| second_level.get(index).map(|value| (value, index)))
+        {
             if existing.deployment_slot == entry.deployment_slot
                 && existing.effective_slot == entry.effective_slot
             {
-                return (true, existing.clone());
+                if matches!(existing.program, LoadedProgramType::Unloaded) {
+                    // The unloaded program is getting reloaded
+                    // Copy over the usage counter to the new entry
+                    entry.usage_counter.store(
+                        existing.usage_counter.load(Ordering::Relaxed),
+                        Ordering::Relaxed,
+                    );
+                    second_level.swap_remove(entry_index);
```

Comment:
for the record: this newly-added instance of `.swap_remove()` is fixed at #32973